CODEWALK.md

```diff
@@ -346,6 +346,118 @@ essentially who is storing what data, for what fee and which sectors have been s
 The keystore contains the binary encoded peer key for interacting securely over the network. 
 This data lives in a file at `$HOME/.filecoin/keystore/self`.
 
+## libp2p Networking
+
+As mentioned above, Filecoin relies on [libp2p](https://libp2p.io/) for its
+networking needs. Here we'll explain some of the mechanics that drive the
+network layer, in particular:
+
+1. Node discovery
+2. NAT topology determination
+3. Autorelay services for solving connectivity issues
+
+### Node discovery
+
+We use the libp2p Kademlia DHT bootstrapping mechanism to find Filecoin peers
+to interface with. libp2p maintains a k-bucket routing table, where peers are
+organized according to the _distance metric proposed_ in the [Kademlia
+paper](http://www.scs.stanford.edu/~dm/home/papers/kpos.pdf).
+
+To populate our routing table, and to keep it healthy over time, we traverse
+the network periodically (and once on startup) by performing random walks. We
+kickstart this process by seeding our routing table with a set of trusted
+bootstrap nodes operated by the Filecoin project. This approach will likely
+evolve in the future.
+
+On each round, we generate a random peer ID and query our known DHT peers by
+sending them `FindPeer(id)` messages. Each node returns a list of peers they
+believe to be closer to the search target (based on the Kademlia distance
+metric), along with their addresses.
+
+As replies come in, we connect to new incoming nodes we've discovered, and we
+repeat this process iteratively, until either:
+
+a. we find the peer; highly unlikely scenario, as the chance is 1/2^256 with
+   our peer ID length; or
+b. we yield because a timer runs out; the most probable scenario. This is a
+   healthy timeout to timebox discovery iterations.
+
+Throughout this process, the libp2p stack generates `CONNECTED` events for
+every new connection we establish to a peer, for which `go-filecoin` registers
+a callback that triggers the Filecoin `HELLO` protocol negotiation. If the
+other party responds positively, chain sync with that peer begins.
+
+### NAT topology determination
+
+In a P2P world, one of the main challenges is achieving connectivity when one,
+or both, of the parties is/are behind a Network Address Translation (NAT)
+devices, such as routers. NATs are commonplace in different network types: 
+residential, companies, public hotspots, universities, etc. Their purpose is 
+to enable devices in a private network to share a common public IP address in 
+the global, scarce IPv4 space.
+
+Outbound connections behind a NAT tend to be frictionless, but inbound
+connections are problematic. For a primer on this topic, we recommend reading
+[Peer-to-Peer Communication Across Network Address
+Translators](https://pdos.csail.mit.edu/papers/p2pnat.pdf).
+
+To enable a pair of Filecoin peers behind NATs to communicate with one
+another, we use libp2p relay services: nodes deployed across the globe that
```

Comment:
not sure if our 5 nodes running on one box in aws constitute nodes deployed across the globe, though i guess strictly speaking that is true :)

Reply:
Haha they're across the globe from where you are :P I'll tone down this statement.